src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -4080,62 +4101,128 @@ EqualOpExpressionLists(List *firstOpExpressionList, List *secondOpExpressionList
  * WorkerLimitCount checks if the given extended node contains a limit node, and
  * if that node can be pushed down. For this, the function checks if this limit
  * count or a meaningful approximation of it can be pushed down to worker nodes.
- * If they can, the function returns the limit count. Otherwise, the function
+ * If they can, the function returns the limit count.
+ *
+ * The limit push-down decision tree is as follows:
+ *                                 group by?
+ *                              1/           \0
+ *                          order by?         (exact pd)
+ *                       1/           \0
+ *           has order by agg?           has aggregate
+ *            1/           \0            1/           \0
+ *     can approximage?    (exact pd)   (no pd)     (exact pd)
+ *      1/       \0   
+ * (approx pd)   (no pd)
+ * 
+ * When an offset is present, the offset value is added to limit because for a query
+ * with LIMIT x OFFSET y, (x+y) records should be pulled from the workers.
+ *
+ * If no limit or offset is present or can be pushed down, then WorkerLimitCount
  * returns null.
  */
 static Node *
 WorkerLimitCount(MultiExtendedOp *originalOpNode)
 {
-	Node *workerLimitCount = NULL;
+	Node *workerLimitNode = NULL;
 	List *groupClauseList = originalOpNode->groupClauseList;
 	List *sortClauseList = originalOpNode->sortClauseList;
 	List *targetList = originalOpNode->targetList;
+	bool hasOrderByAggregate = HasOrderByAggregate(sortClauseList, targetList);
+	bool canApproximate = CanPushDownLimitApproximate(sortClauseList, targetList);
+	bool hasAggregateTarget = HasAggregateTarget(targetList);
+	bool canPushDownLimit = false;
+	bool hasAggregateWithNoSort = groupClauseList != NIL && hasAggregateTarget && 
```

Comment:
Also naming of hasAggregateWithNoSort only implies (hasAggregateTarget && sortClauseList == NIL).